### PR TITLE
[WIP] new listener filter for terminating HTTP/1 CONNECT(#19077)

### PR DESCRIPTION
**Pull request by liverbirdkte**
Add a new listener filter to handle CONNECT request, only HTTP/1.x is supported.

1. inspect if incoming request is HTTP CONNECT

2. drain CONNECT request data out of queue

3. terminate HTTP CONNECT by writing a response to the client

4. set server names using request url in CONNECT request

It seems the behavior of draining CONNECT data and writing response may have an impact on other listener filters in the loop. Is there any way to exit the listener filter iteration? Or I should put this listener filter as the last one in the config.  

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            cb_->socket().ioHandle().resetFileEvents();
+            // Do not skip following listener filters.
+            cb_->continueFilterChain(true);
+            break;
+          case ParseState::Continue:
+            // do nothing but wait for the next event
+            break;
+          }
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.cc:81]: BTW, should we handle a peer closing connection?

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    config_->stats().read_error_.inc();
+    return ParseState::Error;
+  }
+
+  // Remote closed
+  if (result.return_value_ == 0) {
+    return ParseState::Error;
+  }
+  return parseConnect(absl::string_view(reinterpret_cast<const char*>(buf_)));
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.cc:102]: nit: no need to instantiate a string_view here.

**Review by rojkov — COMMENTED** [review]
Nice! I added a few comments.

Also this PR needs docs and tests.

File: api/envoy/extensions/filters/listener/connect_handler/v3/connect_handler.proto [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Connect Handler Filter]
+// Detect and terminate HTTP/1 CONNECT request.
+// [#extension: envoy.filters.listener.connect_handler]
+
+message ConnectHandler {
+  option (udpa.annotations.versioning).previous_message_type =
+      "envoy.config.filter.listener.http_inspector.v2.ConnectHandler";
```
Comment (adisuissa) [inline api/envoy/extensions/filters/listener/connect_handler/v3/connect_handler.proto:19]: I don't think such a type exists.
If there is no v2 predecessor, you should remove the annotation.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this, left an API comment.

**Comment by liverbirdkte** [comment]
Sorry for the late response, thanks for the comments, I'll submit a new version soon. The patch is for issue https://github.com/envoyproxy/envoy/issues/19077.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {
+  http_parser_init(&parser_, HTTP_REQUEST);
+}
+
+http_parser_settings Filter::settings_{
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+static absl::string_view response_body = " 200 Connection Established\r\n\r\n";
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:33]: got it, will use anonymous namespace instead.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc (hunk 1 above)
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:81]: The original listener filter code did register both read and closed events, but closed event is removed due to an issue on Windows, please see https://github.com/envoyproxy/envoy/pull/18265

File: source/extensions/filters/listener/connect_handler/connect_handler.cc (hunk 2 above)
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:102]: got it

File: api/envoy/extensions/filters/listener/connect_handler/v3/connect_handler.proto (hunk 3 above)
Comment (liverbirdkte) [reply api/envoy/extensions/filters/listener/connect_handler/v3/connect_handler.proto:19]: sure, will remove this.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            break;
+          case ParseState::Continue:
+            // do nothing but wait for the next event
+            break;
+          }
+        },
+        Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
+    return Network::FilterStatus::StopIteration;
+  }
+  NOT_REACHED_GCOVR_EXCL_LINE;
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:84]: Do you mean adding a default for switch to handle missing cases or to extend ParseState by adding a new status？

[the remark this answers is not in the document]
Comment (rojkov) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:84]: No need to add `default` as some compilers can generate a warning at compile time about non-exhaustive switch'es.

If `ParseState` is ever extended with a new value then this `switch` will become non-exhaustive. So, it'll be possible to reach `NOT_REACHED_GCOVR_EXCL_LINE`. And that would be a bug most probably. Unfortunately reaching `NOT_REACHED_GCOVR_EXCL_LINE` doesn't trigger a bug nor prints a warning. Thus it's better to replace it with `ENVOY_BUG()`.

**Comment by liverbirdkte** [comment]
start working on tests and docs, will update when code is ready

**Comment by liverbirdkte** [comment]
Add unittest, looking at how to write fuzz tests..

**Comment by rojkov** [comment]
Please never use git force-push. If the PR is not ready for review you can turn it into a Draft.

**Comment by liverbirdkte** [comment]
OK, I was fixing some CI failures, saw the DOC issue, so I added sign-off and force push as the instruction on the page told me to.

Also didn't mean to modify unrelated jwt_authn proto, but git hook format checker won't get passed unless I apply these changes.

**Comment by liverbirdkte** [comment]
Thanks, I will convert it back when it's ready.

File: CODEOWNERS [hunk 6]
```diff
@@ -216,3 +216,6 @@ extensions/filters/http/oauth2 @rgs1 @derekargueta @snowp
 /contrib/sip_proxy/ @durd07 @nearbyfly @dorisd0102
 /contrib/cryptomb/ @rojkov @ipuustin
 /contrib/vcl/ @florincoras @rojkov
+
+# connect handler
+/*/extensions/filters/listener/connect_handler @rojkov @alyssawilk
```
Comment (adisuissa) [inline CODEOWNERS:221]: Please move up to the extensions part.

**Review by adisuissa — COMMENTED** [review]
/lgtm api
left a minor comment.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    }
+    config_->stats().read_error_.inc();
+    return ParseState::Error;
+  }
+
+  // Remote closed
+  if (result.return_value_ == 0) {
+    return ParseState::Error;
+  }
+  return parseConnect(absl::string_view(reinterpret_cast<const char*>(buf_), result.return_value_));
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.cc:106]: No need to instantiate a temp string_view here.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            protocol_ = Http::Headers::get().ProtocolStrings.Http11String;
+          }
+        } else {
+          ENVOY_LOG(trace, "connect:handler: unsupported http version: {}.{}", parser_.http_major,
+                    parser_.http_minor);
+          config_->stats().connect_not_found_.inc();
+          return ParseState::Done;
+        }
+
+        std::vector<absl::string_view> fields = absl::StrSplit(data, absl::ByAnyChar(" :"));
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.cc:142]: Would it be more efficient to use `StringUtil::cropLeft()` and `StringUtil::cropRight()` for server name extraction?

File: source/extensions/filters/listener/connect_handler/connect_handler.h [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Global configuration for CONNECT handler.
+ */
+class Config {
+public:
+  Config(Stats::Scope& scope);
+
+  const ConnectHandlerStats& stats() const { return stats_; }
+
+  static constexpr uint32_t MAX_INSPECT_SIZE = 8192;
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.h:52]: What's the point of having it as a part of public interface? Could it be just a static storage variable?

File: tools/spelling/spelling_dictionary.txt [hunk 10]
```diff
@@ -756,6 +756,7 @@ interpretable
 intra
 ints
 invariance
+iov
```
Comment (rojkov) [inline tools/spelling/spelling_dictionary.txt:759]: This is probably not needed.

**Review by rojkov — COMMENTED** [review]
Thank you! Looks good. I added a few questions.

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {
+  http_parser_init(&parser_, HTTP_REQUEST);
+}
+
+http_parser_settings Filter::settings_{
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:32]: I'm not familiar with Envoy::ThreadLocal, thought it was used for synchronization between main thread and workers, I'll take a look at this.

File: CODEOWNERS (hunk 6 above)
Comment (liverbirdkte) [reply CODEOWNERS]: done

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        if (!result.ok()) {
+          config_->stats().read_error_.inc();
+          return ParseState::Error;
+        }
+        // Remote closed
+        if (result.return_value_ == 0) {
+          return ParseState::Error;
+        }
+        // terminate CONNECT request
+        resp_buf_.add(protocol_);
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc]: done

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_CONNECT_HANDLER_STATS(POOL_COUNTER_PREFIX(scope, "connect_handler."))} {}
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {
+  http_parser_init(&parser_, HTTP_REQUEST);
+}
+
+http_parser_settings Filter::settings_{
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc]: done

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+namespace {
+constexpr absl::string_view response_body = " 200 Connection Established\r\n\r\n";
```
[the remark this answers is not in the document]
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc]: Done

File: source/extensions/filters/listener/connect_handler/connect_handler.cc (hunk 7 above)
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc]: done

File: source/extensions/filters/listener/connect_handler/connect_handler.cc (hunk 8 above)
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc]: done

File: source/extensions/filters/listener/connect_handler/connect_handler.h (hunk 9 above)
Comment (liverbirdkte) [reply source/extensions/filters/listener/connect_handler/connect_handler.h:53]: I followed the patther of http and tls inspectors, are you suggesting setting MAX_INSPECT_SIZE as a static variable in connect_handler.h file?

File: source/extensions/filters/listener/connect_handler/connect_handler.cc [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+namespace {
+constexpr absl::string_view ResponseBody = " 200 Connection Established\r\n\r\n";
+constexpr http_parser_settings Settings{
+    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr,
+};
+}
```
Comment (rojkov) [inline source/extensions/filters/listener/connect_handler/connect_handler.cc:35]: nit:
```suggestion
} // namespace
```

File: source/extensions/filters/listener/connect_handler/connect_handler.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (rojkov) [reply source/extensions/filters/listener/connect_handler/connect_handler.cc:28]: Um, right. Now I'm not sure wrapping this raw buffer in ThreadLocalObject is a good idea. Never mind then.

File: source/extensions/filters/listener/connect_handler/connect_handler.h (hunk 9 above)
Comment (rojkov) [reply source/extensions/filters/listener/connect_handler/connect_handler.h:53]: It could be a constexpr in the anonymous namespace of connect_handler.cc. `static` is redundant in this case.

But now I see it's used in the declaration of `Filter`... Let's have it unchanged then.

File: test/extensions/filters/listener/connect_handler/BUILD [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+licenses(["notice"])  # Apache 2
+
+envoy_package()
+
+envoy_extension_cc_test(
+    name = "connect_handler_test",
+    srcs = ["connect_handler_test.cc"],
+    extension_names = ["envoy.filters.listener.connect_handler"],
+    #TODO(davinci26): The test passes on Windows *but* http inspector
+    # *used* to rely on Event::FileTriggerType::Edge and we got away with it
```
Comment (rojkov) [inline test/extensions/filters/listener/connect_handler/BUILD:20]: Is this comment relevant to connect_handler?

File: test/extensions/filters/listener/connect_handler/BUILD [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+envoy_cc_fuzz_test(
+    name = "connect_handler_fuzz_test",
+    srcs = ["connect_handler_fuzz_test.cc"],
+    corpus = "connect_handler_corpus",
+    deps = [
+        "//source/extensions/filters/listener/connect_handler:connect_handler_lib",
+        "//test/extensions/filters/listener/common/fuzz:listener_filter_fuzzer_lib",
+    ],
+)
```
Comment (rojkov) [inline test/extensions/filters/listener/connect_handler/BUILD:57]: This target seems to be missing in the PR.

File: tools/spelling/spelling_dictionary.txt (hunk 10 above)
Comment (rojkov) [reply tools/spelling/spelling_dictionary.txt]: Indeed the check fails. This is weird as I used iov in my PRs and everything was ok. And it's all over the code base. Looks like the code is treated as a comment by the spellchecker. @phlax could you please have a look?

File: docs/root/configuration/listeners/listener_filters/connect_handler.rst [hunk 18]
```diff
@@ -0,0 +1,36 @@
+.. _config_listener_filters_connect_handler:
+
+Connect Handler
+===============
+
+Connect Handler listener filter allows detecting whether the incoming request appears to be HTTP/1.x CONNECT, and if it is CONNECT, it removes the CONNECT by reading data from socket and responding with a 200 status code to indicate a connection to remote server is established. It also extracts remote server url from the CONNECT. This can be used to select a
```
Comment (adisuissa) [inline docs/root/configuration/listeners/listener_filters/connect_handler.rst:6]:
```suggestion
Connect Handler listener filter allows detecting whether the incoming request appears to be HTTP/1.x CONNECT, and if it is CONNECT, it removes the CONNECT by reading data from the socket and responding with a 200 status code to indicate a connection to the remote server is established. It also extracts the remote server url from the CONNECT. This can be used to select a
```

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a small comment on the doc.
/lgtm api
I suggest adding an integration test.
